### ISSUE TITLE
Support Auto Debug Mode

### ISSUE THREAD

**Issue by vaind** [getsentry/sentry-dotnet#3212]
This is the current code:
https://github.com/getsentry/sentry-dotnet/blob/6274944d00826b0442d038f675f4f46ac0fec182/src/Sentry/SentryOptions.cs#L620-L668

What this effectively does, AFAICT, is that we don't print any warnings when `Debug` is not set. That is pretty unexpected behavior, at least to me, especially considering what the docs say when you're setting up sentry:

```
    // When debug is enabled, the Sentry client will emit detailed debugging information to the console.
    // This might be helpful, or might interfere with the normal operation of your application.
    // We enable it here for demonstration purposes when first trying Sentry.
    // You shouldn't do this in your applications unless you're troubleshooting issues with Sentry.
    options.Debug = true;
```

Instead, I'd expect the DiagnosticLogger property to always be reachable. It already has a check which log levels should be printed.

**Comment by bitsandfoxes** [comment]
Looks like `Debug` controls whether the SDK adds its own `ConsoleLogger` as `DiagnosticLogger` if none is present. 
https://github.com/getsentry/sentry-dotnet/blob/dfe208fa3cd51ccdcf0853cbf8b44e2fae2453c3/src/Sentry/SentryOptions.cs#L1636-L1645

> Instead, I'd expect the DiagnosticLogger property to always be reachable. It already has a check which log levels should be printed.

You can set your own logger and that works independently whether `Debug` is enabled or not.

**Comment by vaind** [comment]
Actually, no, you cannot set a diagnostic logger. 

https://github.com/getsentry/sentry-dotnet/blob/6274944d00826b0442d038f675f4f46ac0fec182/src/Sentry/SentryOptions.cs#L651-L653

Can we just remove the need to set Debug for DiagnosticLogger?

**Comment by bitsandfoxes** [comment]
> Can we just remove the need to set Debug for DiagnosticLogger to work at all?

As in: If there is a `Diagnostic`Logger present, log?

**Comment by jamescrosswell** [comment]
We could/should be able to control everything via the `DiagnosticLevel` right?
 https://github.com/getsentry/sentry-dotnet/blob/0711c17f539a1083ca783555db37368267d32176/src/Sentry/SentryOptions.cs#L641

So SentryOptions.Debug is a bit redundant... although lots of customers will definitely know about it so if we do phase it out, probably need to mark it as `Obsolete` (and keep it as a proxy to setting the `DiagnosticLevel`) until some time in the distant future (e.g. v5.0). 

Ideally we'd change the default value of `SentryOptions.DiagnosticLevel` to `Warning` as well... if people did set `Debug = true`, that would be (for the time being) an alternative way of setting  `SentryOptions.DiagnosticLevel = SentryLevel.Debug`

**Comment by vaind** [comment]
> So SentryOptions.Debug is a bit redundant... although lots of customers will definitely know about it so if we do phase it out, probably need to mark it as Obsolete (and keep it as a proxy to setting the DiagnosticLevel) until some time in the distant future (e.g. v5.0).

it is but it's common to all SDKs so I would keep it, just use it to set the diagnostic level to Debug, instead of the default warning - as you're suggesting. The main gripe I have with the current code is the need to enable Debug to get **any logs**, even warnings

**Comment by jamescrosswell** [comment]
> The main gripe I have with the current code is the need to enable Debug to get **any logs**, even warnings

Yeah fair enough. What do you think should happen if people set `SentryOptions.Debug = false` though (which is the default value for that property)? Would that set the DiagnosticLevel to `Error`?

If we leave SentryOptions.Debug in the SDK, we have to deal with 10 different combinations of those two settings (Debug x DiagnosticLevel) as well as the order that these are set in (if people are using the Options callback) and which one takes precedence (if people are just passing in an options object). Not saying that can't be done but we'd need to decide how it should work first right?

We could do it a little bit like EnableTracing, which implicitly sets TraceSampleRate to 1.0 or 0.0 depending on whether it's set to true or false. EnableTracing is a `bool?` though (not a `bool`). So maybe we could change `SentryOptions.Debug` to be a `bool?` and do something similar there? If it's true, we set `SentryOptions.DiagnosticLevel` to Debug... if it's false we set it to `Error`... or alternatively people can get more granular by setting `SentryOptions.DiagnosticLevel` directly. If they've set both though, which one takes precedence?

**Comment by vaind** [comment]
This came up again and there're now develop docs describing how this should be handled.

https://develop.sentry.dev/sdk/expected-features/#debug-mode

**Comment by jamescrosswell** [comment]
> https://develop.sentry.dev/sdk/expected-features/#debug-mode

Those docs include this:

```
Every SDK must implement a debug mode, which is disabled by default
```

... as long as that's the case, I can't see how we can log warnings/errors by default. Would really need to argue for a change to the default behaviour recommended/implemented by all SDKs.

**Comment by vaind** [comment]
Can we do this?

> [Auto Debug Mode](https://develop.sentry.dev/sdk/expected-features/#auto-debug-mode)
> So users can easily spot errors during development, the SDK can automatically enable debug mode when it reliably detects it's running in a debug build. When doing so, the SDK must communicate this in the docs and with a log message when initializing the SDK. When the user explicitly sets the debug option to false, the SDK must disable the debug mode.

**Comment by Flash0ver** [comment]
Discussion with @jamescrosswell 
- We could provide a Source-Generated default value, particularly for NativeAOT, that utilizes
  - the `$(Configuration)` property
  - `CSharpParseOptions.PreprocessorSymbolNames`
  - `CompilationOptions.OptimizationLevel`
  - emit `#if DEBUG .. #else .. #endif` C# into the consumer's compilation
- and as a fallback, if not source generated
  - `Debugger.IsAttached`
  - Reflection: `[assembly: DebuggableAttribute]`

**Comment by Flash0ver** [comment]
Discussion with @jamescrosswell:
- we could have `Auto Debug Mode` through Source-Generation
- however, our default level is `DiagnosticLevel = SentryLevel.Debug`
  - which is quite noisy
  - and - e.g. in a Console-App - difficult to spot the user code's messages
- we could do something like
  - if not explicitly set, but detected, set the `DiagnosticLevel = SentryLevel.Error/Fatal`
- but we don't see the immediate benefit and there was no user ask so far

Another question arising: do we always, also on `Release`, log Errors/Fatal?
This may lead to a bit of a refactoring, but then the behavior between the Log-Levels is not consistent.

We ended up agreeing with leaving it as is.
We saw user code that needs to deal with this inconvenience via
```CSharp
#if DEBUG
options.Debug = true;
#endif
```

We're closing this issue as we don't intend to support `Auto Debug Mode` in the next major release.
Also, no need to update documentation, as only SDKs that have this feature must actively document it.

**Comment by vaind** [comment]
> however, our default level is DiagnosticLevel = SentryLevel.Debug

Just to clarify - some SDKs, for the same reason, changed the default log level

---

As is, I still believe the SDK should be printing warnings and errors by default. They provide important information and **visibility** to issues at runtime (for example, not sending any events to Sentry because of misconfiguration)

**Comment by Flash0ver** [comment]
> As is, I still believe the SDK should be printing warnings and errors by default. They provide important information and **visibility** to issues at runtime (for example, not sending any events to Sentry because of misconfiguration)

Hmmm ... well ... perhaps there is a way we can figure out for this to be helpful, yet not too noisy, as discussed in #4330.

Is this what you have in mind? (also considering the dev docs)
- when explicit `SentryOptions.Debug = true`
  - behavior as is, with `DiagnosticLevel = SentryLevel.Debug` per default
  - for all Build Configurations
- when explicit `SentryOptions.Debug = false`
  - suppress "Auto Debug Mode"
- when `'$(Configuration)'=='DEBUG'`, and `SentryOptions.Debug` is not explicitly set
  - use a Diagnostic-Logger with either `SentryLevel.Warning` or `SentryLevel.Error` or `SentryLevel.Fatal` as minimum level
  - detectable via a Source-Generator

A couple of thoughts:
- our current Source-Generator approach depends on `[ModuleInitializer]` and only works for `net5.0`
  - we would see different "Auto Debug Mode" behavior between .NET Framework and .NET apps (that have not reached End-Of-Support yet)
- our current Source-Generator only supports C#
  - but we could expand it to also work for Visual Basic projects
  - however, we would need a different solution for F#, as it's not based on the Roslyn Compiler
- we could instead have a Source-Generator that works with C# Interceptors instead
  - but this requires a `.NET 9`-compatible version of the Compiler
  - and only works for C#, no support for Visual Basic
  - and again, no F#

Which leads me to a question:
- a Source-Generator-based approach would not work for `sentry-unity` nor `sentry-powershell`
  - `sentry-unity`: I guess in Unity we could inject something similar to the consuming game, to detect a Debug-Configuration, couldn't we @bitsandfoxes?
  - `sentry-powershell`: for PowerShell I have never thought about the difference of a Debug-vs-Release build
    - @vaind is there some facility to detect the consuming script is in a Debug-Configuration/Debug-Mode
    - and is there some facility for the Sentry SDK to detect that and configure the Diagnsotic-Logger accordingly

TL;DR;
Questions about behavior between older .NET Target Frameworks and newer .NET Target Frameworks;
and between C#, VB and F#;
and between `sentry-dotnet`, `sentry-powershell` and `sentry-unity`.

**Comment by vaind** [comment]
It is perfectly fine to have things only work for a subset of SDK users, if the subset is reasonably large. I don't know how many VB or F# consumers we have (I am not aware whether we track this) but I'm quite convinced it's safe to assume most would be using c#. If we can improve developer-experience for 90+ % of new users with a reasonable effort, I believe we should do that, but you're the maintainer of the SDK so it's up to you.

- Unity: I think we could easily implement auto-debug mode over in the Unity because there're APIs to check whether we run in Editor and whether the game build is a debug build.
- PowerShell - we already have it https://github.com/getsentry/sentry-powershell/blob/9a1d9ba8dd719c58a97bbdec8e1898966eed8fd3/modules/Sentry/public/Start-Sentry.ps1#L28-L37

**Comment by Flash0ver** [comment]
@jamescrosswell and I reiterated, and we're thinking about updating a prominent sample, or maybe most/all of the samples, in the docs with
```
#if DEBUG
options.Debug = true;
#endif
```
